src/server/pfs/server/paths.go

```diff
@@ -0,0 +1,76 @@
+package server
+
+import (
+	"path"
+	"regexp"
+	"strings"
+
+	globlib "github.com/pachyderm/ohmyglob"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
+)
+
+var globRegex = regexp.MustCompile(`[*?[\]{}!()@+^]`)
+
+func globLiteralPrefix(glob string) string {
+	idx := globRegex.FindStringIndex(glob)
+	if idx == nil {
+		return glob
+	}
+	return glob[:idx[0]]
+}
+
+func matchFunc(glob string) (func(string) bool, error) {
+	// TODO Need to do a review of the globbing functionality.
+	var parentG *globlib.Glob
+	parentGlob, baseGlob := path.Split(glob)
+	if len(baseGlob) > 0 {
+		var err error
+		parentG, err = globlib.Compile(parentGlob, '/')
+		if err != nil {
+			return nil, err
+		}
+	}
+	g, err := globlib.Compile(glob, '/')
+	if err != nil {
+		return nil, err
+	}
+	return func(s string) bool {
+		s = path.Clean(s)
+		return g.Match(s) && (parentG == nil || !parentG.Match(s))
+	}, nil
+}
+
+func parseGlob(glob string) (index.Option, func(string) bool, error) {
+	mf, err := matchFunc(glob)
+	if err != nil {
+		return nil, nil, err
+	}
+	prefix := globLiteralPrefix(glob)
+	return index.WithPrefix(prefix), mf, nil
+}
+
+func pathIsChild(parent, child string) bool {
+	parent = strings.TrimLeft(parent, "/")
+	child = strings.TrimLeft(child, "/")
+	if !strings.HasPrefix(child, parent) {
+		return false
+	}
+	rel := child[len(parent):]
+	rel = strings.Trim(rel, "/")
+	return !strings.Contains(rel, "/")
+}
+
+// cleanPath converts paths to a canonical form used in the driver
+// "" -> "/"
+// "abc" -> "/abc"
+// "/abc" -> "/abc"
+// "/" -> "/"
+func cleanPath(x string) string {
+	if x == "" {
+		x = "/"
+	}
+	if len(x) > 1 && x[0] != '/' {
```

Comment:
I think this is going to miss a length 1 path? I don't think you need the length check. I think this could just be:
```
if strings.HasPrefix(str, "/") {
  return str
}
return "/" + str
```